### PR TITLE
Story/fireweather bot/367

### PR DESCRIPTION
**Pull request by andrea-williams**
* Alembic migration script to add new noon_forecasts table to DB
* fireweather_bot script to post queries to BC FireWeather Phase 1 API for all weather stations, downloading CSV file of results, parse CSV and save it to DB
* Openshift template to add a cronjob for fireweather_bot so it runs daily at 8:30 am and 4:30 pm

### CODE REVIEW

File: app/fireweather_bot.py [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        json_data = json.load(weather_stations_file)
+        return json_data['weather_stations']
+
+
+def _get_station_names_to_codes() -> Dict:
+    """ Helper function to create dictionary of (station_name: station_code) key-value pairs
+    Is used when replacing station names with station IDs in dataframe
+    """
+    station_data = _get_stations_local()
+    station_codes = {}
```
Comment (Sybrand) [inline app/fireweather_bot.py:240]: Sure - that works - but how sexy is this: `station_codes = { station['name'] : station['code'] for station in station_data}`

File: app/main.py [hunk 2]
```diff
@@ -119,6 +120,18 @@ async def get_forecast_summaries(model: ModelEnum, request: schemas.StationCodeL
         LOGGER.critical(exception, exc_info=True)
         raise
 
+@app.post('/noon_forecasts/', response_model=schemas.NoonForecastResponse)
+def get_noon_forecasts(request: schemas.StationCodeList, _: bool = Depends(authenticate)):
+    """ Returns noon forecasts pulled from BC FireWeather Phase 1 website for the specified
+    set of weather stations. """
+    try:
+        start_date = datetime.datetime.now().replace(hour=0, minute=0, second=0)
```
Comment (Sybrand) [inline app/main.py:128]: Please throw in a LOGGER.info('/noon_forecasts/') right after the try - I'm trying to have that on every api call.

File: app/main.py [hunk 3]
```diff
@@ -119,6 +120,18 @@ async def get_forecast_summaries(model: ModelEnum, request: schemas.StationCodeL
...
+        end_date = start_date + datetime.timedelta(days=5)
+        LOGGER.info('Querying /noon_forecasts/ for %s from %s to %s', request.stations, start_date, end_date)
+        return fetch_noon_forecasts(request.stations, start_date, end_date)
+    except Exception as exception:
+        LOGGER.critical(exception, exc_info=True)
```
Comment (Sybrand) [inline app/main.py:133]: We really need to add a catch-all exception in - I'll make a ticket for it....

File: app/noon_forecasts.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+
+def parse_table_records_to_noon_forecast_response(data: [NoonForecasts]):
+    """ Given a list of table records from the database, parse each record
+    (which is a NoonForecasts object) and structure it as a NoonForecast
+    object, then return the list of NoonForecast objects as a NoonForecastResponse
+    """
+    noon_forecasts = {}
+    # offset is used to convert Pacific Daylight time (noon) to UTC
+    UTC_OFFSET = +7
```
Comment (Sybrand) [inline app/noon_forecasts.py:29]: Please remove - the database should now be storing the timestamp with a timezone, so you don't have to do anything here - the offset should just be there.

File: app/noon_forecasts.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    for record in data:
+        station_code = record.station_code
+        noon_forecast_value = NoonForecastValue(
+            datetime=datetime.datetime(
+                year=record.weather_date.year,
+                month=record.weather_date.month,
+                day=record.weather_date.day,
+                hour=13 + UTC_OFFSET,
```
Comment (Sybrand) [inline app/noon_forecasts.py:37]: Maybe I'm just confused - but I would expect the object in the database to already have the correct time and timezone.

File: app/noon_forecasts.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            temperature=record.temperature,
+            rh_valid=record.rh_valid,
+            relative_humidity=record.relative_humidity,
+            wdir_valid=record.wdir_valid,
+            wind_direction=record.wind_direction,
+            wspeed_valid=record.wspeed_valid,
+            wind_speed=record.wind_speed,
+            precip_valid=record.precip_valid,
+            total_precipitation=record.precipitation,
+            gc=None if math.isnan(record.gc) else record.gc,
```
Comment (Sybrand) [inline app/noon_forecasts.py:49]: sonarqube complains about this way of nan checking - but I'm fine overriding it unless you think of some genius method.

File: app/noon_forecasts.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    """ Given a list of table records from the database, parse each record
+    (which is a NoonForecasts object) and structure it as a NoonForecast
+    object, then return the list of NoonForecast objects as a NoonForecastResponse
+    """
+    noon_forecasts = {}
+    # offset is used to convert Pacific Daylight time (noon) to UTC
+    UTC_OFFSET = +7
+    for record in data:
+        station_code = record.station_code
+        noon_forecast_value = NoonForecastValue(
```
Comment (Sybrand) [inline app/noon_forecasts.py:32]: It still think we can do NoonForecastValue(**data) here - and just tinker with the data a bit - but at this point I don't care enough.

File: openshift/templates/bcfw_p1_forecasts.cronjob.yaml [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+          labels:
+            cronjob: ${JOB_NAME}
+        spec:
+          template:
+            spec:
+              containers:
+                - name: ${JOB_NAME}
+                  image: docker-registry.default.svc:5000/${PROJECT_TOOLS}/${NAME}-${SUFFIX}:${SUFFIX}
+                  imagePullPolicy: "Always"
+                  command: ["python", "fireweather_bot.py"]
```
Comment (Sybrand) [inline openshift/templates/bcfw_p1_forecasts.cronjob.yaml:43]: Should be
```suggestion
                  command: ["python", "-m", "app.fireweather_bot"]
```

File: pyproject.toml [hunk 9]
```diff
@@ -21,6 +21,9 @@ uvicorn = "^0.11.5"
 geopandas = "^0.7.0"
 pylint = "^2.5.3"
 alembic = "^1.4.2"
+pytz = "^2020.1"
```
Comment (Sybrand) [inline pyproject.toml:24]: Can we please try not include pytz? I think everything can be done using the built in datatime.timezone

File: pyproject.toml [hunk 10]
```diff
@@ -21,6 +21,9 @@ uvicorn = "^0.11.5"
...
+requests_ntlm = "^1.1.0"
+python-dotenv = "^0.14.0"
```
Comment (Sybrand) [inline pyproject.toml:26]: Is python-dotenv being used anywhere? Can we remove it?

File: app/fireweather_bot.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+    # write the data_df to the database one row at a time, so that if data_df contains >=1 rows that are
+    # duplicates of what is already in the db, the write won't fail for the unique rows
+    # NOTE: iterating over the data_df one Series (row) at a time is ugly, but until pandas is updated
+    # with a fix, this is the easiest work-around.
+    # See https://github.com/pandas-dev/pandas/issues/15988
+    for index, row in data_df.iterrows():
+        try:
+            row = row.dropna()
+            session.add(NoonForecasts(**row.to_dict()))
+            session.flush()
```
[the remark this answers is not in the document]
Comment (andrea-williams) [reply app/fireweather_bot.py:195]: flush() is only called with commit() if autoflush=True on the session, which it isn't (in app.db.database). Although I'm not aware of any reason why we wouldn't use autoflush=True for the session?

File: pyproject.toml (hunk 10 above)
Comment (andrea-williams) [reply pyproject.toml:26]: I've removed python-dotenv from pyproject.toml, but it's still in poetry.lock as a package.extra on the pydantic package from a couple months ago

File: pyproject.toml (hunk 9 above)
Comment (andrea-williams) [reply pyproject.toml:24]: I'm not sure why this is showing up as a change I made, but pytz is being used in a bunch of places (env_canada, spotwx, test_get_hourlies, etc.). Getting rid of pytz is a larger refactor job than the scope of this ticket

File: app/fireweather_bot.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Sybrand) [reply app/fireweather_bot.py:193]: Not so: https://docs.sqlalchemy.org/en/13/orm/session_basics.html#session-committing

"Session.commit() is used to commit the current transaction. It always issues Session.flush() beforehand to flush any remaining state to the database; this is independent of the “autoflush” setting."

File: pyproject.toml (hunk 9 above)
Comment (Sybrand) [reply pyproject.toml:24]: Fair enough - I've gone ahead an done that re-factor on my branch - so we'll get rid of it over time.

File: app/db/models.py [hunk 12]
```diff
@@ -115,3 +117,54 @@ class ModelRunGridSubsetPrediction(Base):
...
+    temperature = Column(Float, nullable=False)
+    rh_valid = Column(Boolean, default=False, nullable=False)
+    relative_humidity = Column(Float, nullable=False)
+    wdir_valid = Column(Boolean, default=False, nullable=False)
+    wind_direction = Column(Float, nullable=False)
+    wspeed_valid = Column(Boolean, default=False, nullable=False)
+    wind_speed = Column(Float, nullable=False)
+    precip_valid = Column(Boolean, default=False, nullable=False)
+    precipitation = Column(Float, nullable=False)
+    gc = Column(Float, nullable=False, default=math.nan)
```
[the remark this answers is not in the document]
Comment (Kyubinhan) [reply app/db/models.py:166]: Is this somehow resolved behind the scene?

File: app/db/database.py [hunk 13]
```diff
@@ -8,7 +8,7 @@
 DB_STRING = 'postgres://{}:{}@{}:{}/{}'.format(
     config.get('POSTGRES_USER', 'wps'),
     config.get('POSTGRES_PASSWORD', 'wps'),
-    config.get('POSTGRES_HOST', 'db'),
+    config.get('POSTGRES_HOST', 'localhost'),
```
Comment (Kyubinhan) [inline app/db/database.py:11]: This will become `config.get('POSTGRES_HOST', 'db'),` once you pull the main branch (merged with Sy's)

File: app/fireweather_bot.py [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+from abc import abstractmethod, ABC
+from requests import Session
+from requests.compat import urljoin
+from requests_ntlm import HttpNtlmAuth
+from sqlalchemy.exc import IntegrityError
+import pandas as pd
+from app import config
+import app.db.database
+from app.db.models import NoonForecasts
+from app.wildfire_one import _get_now, _get_stations_local
```
Comment (Sybrand) [inline app/fireweather_bot.py:22]: Convention is for underscore on functions to indicate it's to be used internally only - but whatever - point is - we should roll this out into a utils file at some point and rename that _get_now to something that indicates that it's utc at some point. (Not part of this PR)

File: openshift/templates/bcfw_p1_forecasts.cronjob.yaml [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+objects:
+  - kind: CronJob
+    apiVersion: batch/v1beta1
+    metadata:
+      name: ${JOB_NAME}
+      labels:
+        cronjob: ${JOB_NAME}
+    spec:
+      schedule: "30 8,16 * * *"
+      concurrencyPolicy: "Forbid"
```
Comment (Sybrand) [inline openshift/templates/bcfw_p1_forecasts.cronjob.yaml:31]: Just an f.y.i. - keep and eye on your cronjob. The env Canada one got stuck, and never stopped, so other jobs never started up. Which is good and bad - just monitor this job for a while to make sure it doesn't have similar issues.

Comment (Kyubinhan) [reply openshift/templates/bcfw_p1_forecasts.cronjob.yaml:31]: Interesting. Maybe we can go with `"Replace"` instead?

Comment (Sybrand) [reply openshift/templates/bcfw_p1_forecasts.cronjob.yaml:31]: I like that - yeah - let's go with Replace here instead of Forbid!

Comment (Sybrand) [reply openshift/templates/bcfw_p1_forecasts.cronjob.yaml:31]:
```suggestion
      concurrencyPolicy: "Replace"
```

File: app/main.py [hunk 16]
```diff
@@ -114,6 +115,19 @@ async def get_model_forecast_summaries(
         LOGGER.critical(exception, exc_info=True)
         raise
 
+@app.post('/noon_forecasts/', response_model=schemas.NoonForecastResponse)
+def get_noon_forecasts(request: schemas.StationCodeList, _: bool = Depends(authenticate)):
+    """ Returns noon forecasts pulled from BC FireWeather Phase 1 website for the specified
+    set of weather stations. """
+    try:
+        LOGGER.info('/noon_forecasts/')
+        start_date = datetime.datetime.now().replace(hour=0, minute=0, second=0)
```
Comment (Kyubinhan) [inline app/main.py:124]: I think start date needs to be `datetime.datetime.now(tz=datetime.timezone.utc)` now that the graph can actually show the reference line with the current time
